### PR TITLE
tgui Preferences Menu + total rewrite of the preferences backend

### LINKED ISSUE #28907: Preferences that are set to enable by default don't enable by default.
For example, take my tooltips PR. in preferences.dm the var that enables tooltips is set to TRUE (`var/enable_tips = TRUE`) but ingame it starts off disabled. Something is disabling it. This same problem affects my reverb PR

### LINKED ISSUE #44887: OOC perfs arent saving
Issue reported from  Round ID: 111980 (/tg/Station Terry [EU] [100% LAG FREE] [New-Dedi])

Reporting client version: 512.1471

OOC preferences are not saving like lobby music and tthe auto deadmin

### LINKED ISSUE #58458: Character setup initially displays the player model in only one direction
## Round ID:
Tested locally

## Testmerges:
N/A

## Reproduction:
Boot it up
Open char prefs
Don't change anything
Get ass
image

### LINKED ISSUE #59181: Cannot Disable Ambient Occlusion
Reporting client version: 513.1539

## Round ID:
[162461](https://scrubby.melonmesa.com/round/162461)

## Testmerges:
- [Fixes sound loops not well, stopping. ](https://www.github.com/tgstation/tgstation/pull/59138)
- [Turns lighting objects into a datum, makes all lighting be performed with an underlay. big maptick fix very good!](https://www.github.com/tgstation/tgstation/pull/58991)

## Reproduction:
Attempt to disable Ambient Occlusion

### LINKED ISSUE #59912: Preferences tab fails to fully load in the lobby
## Reproduction:

Happens on my machine v514.1557 and server also v514.1557

1. Join a prod server and click the Preferences tab:
image
2. Get frustrated when your prefs get wiped and you need to click through 3 screens just to `Stop Sounds` the lobby music
3. Load up a local server to try and solve the issue only to be unable to reproduce it even with localhost admin disabled:
image
4. Cry

### LINKED ISSUE #60823: GLOB.roundstart_races is only populated when SSjob is initialized, so appearance randomization runtimes if done before then
## Reproduction:

For some reason the list of config enabled roundstart races is populated by Initialize in SSjob (`generate_selectable_species`)
https://github.com/tgstation/tgstation/blob/badc431a9c9a27c351b5d4e70826eaa8dee6f2f8/code/controllers/subsystem/job.dm#L71-L79

Meaning, if a fresh player connects before SSjobs is done initializing, this bit of randomization runtime errors, as it is trying to pick from an empty list. This can happen semi-often since `/datum/preferences/New()` calls `randomise_appearance_prefs`, and new preferences are made when fresh players connect (example: guest accounts).
https://github.com/tgstation/tgstation/blob/badc431a9c9a27c351b5d4e70826eaa8dee6f2f8/code/modules/mob/dead/new_player/preferences_setup.dm#L10-L12

image

### Potentially fixed by new preferences menu, leaving this open for tracking purposes.

### PR DESCRIPTION
**Pull request by Mothblocks**
## About The Pull Request

Rewrites the entire preferences menu in tgui. Rewrites the entire backend to be built upon datumized preferences, rather than constant additions to the preferences base datum.

https://user-images.githubusercontent.com/35135081/132632991-fb7d8e73-6b3a-4a7a-9795-b562d665a606.mp4

https://user-images.githubusercontent.com/35135081/132633011-e7d60352-520e-4ec0-b038-6379928137b3.mp4

Splits game preferences into its own window.

Antagonists are now split into their individual rulesets. You can now be a roundstart heretic without signing up for latejoin heretic, as an example.

This iteration matches parity, and provides very little new functionality, but adding anything new will be much easier.

Fixes #60823
Fixes #28907
Fixes #44887
Fixes #59912
Fixes #58458
Fixes #59181

### Major TODOs
- [x] Quirk icons, from @Fikou (with some slight adjustments from me)
- [x] Lore text, from @EOBGames (4/6, need moths and then ethereal lore from @AMonkeyThatCodes)
- [x] Heavy documentation on how one would add new preferences, species, jobs, etc
- [x] A lot of specialized testing so that people's real data don't get corrupted

## Changelog

:cl: Mothblocks, Floyd on lots of the design
refactor: The preferences menu has been completely rewritten in tgui.
refactor: The "Stop Sounds" verb has been moved to OOC.
/:cl:

### CODE REVIEW

**Comment by tralezab** [comment]
ive made bigger prs psssh

**Comment by Iamgoofball** [comment]
░░░░░░░░░░░░▄▄                     
░░░░░░░░░░░█░░█                  
░░░░░░░░░░░█░░█                  
░░░░░░░░░░█░░░█                  
░░░░░░░░░█░░░░█                  
███████▄▄█░░░░░██████▄ 
▓▓▓▓▓▓█░░░░░░░░░░░░░░█ 
▓▓▓▓▓▓█░░░░ /tg/,     ░░░░░█ 
 ▓▓▓▓▓▓█░░░░Add This░░░░█  
▓▓▓▓▓▓█░░░░Please!░░░░░█ 
▓▓▓▓▓▓█░░░░░░░░░░░░░░█ 
▓▓▓▓▓▓█████░░░░░░░░░█   
██████▀░░░░▀▀██████▀

**Comment by tralezab** [comment]
but uuuum you know on first glance looks pretty good 😋

**Comment by Colovorat** [comment]
image

File: code/modules/unit_tests/reagent_id_typos.dm
Comment (Iamgoofball) [inline code/modules/unit_tests/reagent_id_typos.dm:7]: Unintended removal?

Comment (Mothblocks) [reply code/modules/unit_tests/reagent_id_typos.dm:7]: No, chemical reagents not being generated right away was causing issues with organs were being created, because organs now must be created immediately (because we want to generate humans right away), which means that they need to be edible right away, which tries to create nutriments, but nutriments weren't created until after asset generation was complete.

Everything relevant here has been made into a GLOBAL_LIST_INIT.

**Comment by LemonInTheDark** [comment]
Living in the god damn future baby

**Comment by FlamingLily** [comment]
🥳 Let's go!

File: code/modules/client/preferences/assets.dm
Comment (AnturK) [inline code/modules/client/preferences/assets.dm:42]: Waiting for someone to name savefile key `.hat_brim_size.`

**Review by AnturK — COMMENTED** [review]
Looks alright so far. 103/32

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm
Comment (Mothblocks) [reply code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:639]: This is only ever used by prefs menu in order to get the icon, the `/execute()` of this is overridden anyway.

File: code/game/objects/items/inspector.dm
Comment (Mothblocks) [reply code/game/objects/items/inspector.dm:40]: Okay, get ready for this.

`cell` starts out as a type.
In `Initialize`, the cell is replaced by an atom version of itself.
But because we're creating stuff in the assets SS (which was moved *before* atoms, to be as responsive as possible), *and* destroying them, that means we're destroying things before Initialize is called.
There's a few ways around this, such as deferring deletion, but for reasons I don't remember the best solution ended up just being to patch around the few Destroy() calls before Initialize.

File: code/modules/client/preferences/_preference.dm
Comment (Mothblocks) [reply code/modules/client/preferences/_preference.dm:144]: It's not *necessary*, but just makes it so we can just call `is_randomizable()` everywhere else. I could lift the SHOULD_NOT_OVERRIDE bit though.

File: tgui/packages/tgui/components/Box.tsx
Comment (stylemistake) [inline tgui/packages/tgui/components/Box.tsx:260]: This will hit performance very terribly, please don't do it.

File: tgui/packages/tgui/components/Stack.tsx
Comment (stylemistake) [inline tgui/packages/tgui/components/Stack.tsx:14]: Weird indentation

File: tgui/packages/tgui/http.ts
Comment (stylemistake) [inline tgui/packages/tgui/http.ts:1]: Shouldn't import this type from jsdom, which is primarily used for testing. I think the real type is called `RequestInit`.

File: code/modules/client/preferences/_preference.dm
Comment (AnturK) [reply code/modules/client/preferences/_preference.dm:144]: That's fine

File: tgui/packages/tgui/components/Box.tsx
Comment (Mothblocks) [reply tgui/packages/tgui/components/Box.tsx:260]: I think keybindings menu needed this so that we could get refs of the sections to teleport to, I believe because it used `Stack.Item` to get a ref to. Maybe I can use wrapper divs instead.

Comment (stylemistake) [reply tgui/packages/tgui/components/Box.tsx:260]: Just use normal divs, if possible, no need for wrappers.

**Comment by BadIsMyUsername** [comment]
Did a tiny bit of testing, here's some stuff I saw
- Exiting out of the game and then re-entering removes the preference menu verbs from the OOC tab.
- Changing species (or gender?) makes it so you cannot use the boxes to change your clothing/hair/etc.

File: tgui/packages/tgui/components/Box.tsx
Comment (Mothblocks) [reply tgui/packages/tgui/components/Box.tsx:260]: I meant wrapping the content with the div, ya.

File: code/game/machinery/telecomms/broadcasting.dm
Comment (Mothblocks) [reply code/game/machinery/telecomms/broadcasting.dm:173]: CHAT_GHOSTRADIO I meant.

File: code/modules/admin/verbs/anonymousnames.dm
Comment (Mothblocks) [inline code/modules/admin/verbs/anonymousnames.dm:132]: need to fix this name back

File: code/modules/antagonists/changeling/changeling.dm
Comment (Mothblocks) [inline code/modules/antagonists/changeling/changeling.dm:743]: Leftover from a merge conflict

Comment (Fikou) [reply code/modules/antagonists/changeling/changeling.dm:743]: cool

File: code/modules/client/preferences/_preference.dm
Comment (Mothblocks) [inline code/modules/client/preferences/_preference.dm:432]: Need to document

**Comment by GoldenAlpharex** [comment]
Very nice, I can't wait for this to be merged :)

File: code/modules/antagonists/traitor/datum_traitor.dm
Comment (MrMelbert) [inline code/modules/antagonists/traitor/datum_traitor.dm:360]: I don't think these icon_states exist anymore, renamed to `e_sword_on_red`

**Comment by Mothblocks** [comment]
Moved supplementary colors to their own features

https://user-images.githubusercontent.com/35135081/132769556-0bf9d5af-0b5e-44b3-9221-685adeb4228b.mp4

**Comment by Mothblocks** [comment]
2021-09-09T15-31-11
Made species buttons bigger.

**Comment by AMonkeyThatCodes** [comment]
you put ethereals in the wrong spot they should be on top

File: tgui/packages/tgui/components/Box.tsx
Comment (Mothblocks) [reply tgui/packages/tgui/components/Box.tsx:260]: I limited it to Stack.Item only, which uses div.

**Comment by Cyberboss** [comment]
Oh llaaAAAWWWWDDDD~~!!

File: code/modules/antagonists/cult/cult.dm
Comment (Cyberboss) [inline code/modules/antagonists/cult/cult.dm:74]: Eyes tired, I got to here

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/components/Stack.tsx
Comment (Mothblocks) [reply tgui/packages/tgui/components/Stack.tsx]: Should be fixed now

**Comment by Mothblocks** [comment]
BYOND bug, I had zero idea how to fix it. It fixes when you hover over or if you change anything.

File: tgui/packages/tgui/components/Box.tsx
Comment (Mothblocks) [reply tgui/packages/tgui/components/Box.tsx]: I fixed this to `innerRef`

**Comment by Mothblocks** [comment]
Also need to remove the countless unused imports I have.

**Comment by Mothblocks** [comment]
> Changing species (or gender?) makes it so you cannot use the boxes to change your clothing/hair/etc. Only seems to happen some of the time, can't reproduce.

@BadIsMyUsername This was causing me headaches for weeks but I figured it out. When you open a clothing selection, then click on another tab, then come back, it would not open anymore.

The cause is Popper running its own Inferno `render`, but then removing the element directly rather than rendering null. This means `TrackOutsideClick` was never unmounted, which means it was still capturing click calls, which instantly closed the menu.

File: code/modules/antagonists/traitor/datum_traitor.dm
Comment (Mothblocks) [reply code/modules/antagonists/traitor/datum_traitor.dm:360]: image
🎉

File: code/__HELPERS/mobs.dm
Comment (Cyberboss) [inline code/__HELPERS/mobs.dm]: This is officially too woke for me

Comment (necromanceranne) [reply code/__HELPERS/mobs.dm]: They're the real names of those skin tones. I'm also pretty sure this isn't the first time I've seen someone push for a change like this. I can't remember why it wasn't merged.

Comment (stylemistake) [reply code/__HELPERS/mobs.dm]: It's definitely not better, for me beige is no different than sand, and previous color names are much easier to sort through.

If I want to be caucasian-skinned, only three causasian tan levels to choose from, each in increasing strength level.

Comment (Mothblocks) [reply code/__HELPERS/mobs.dm]: It doesn't require color literacy:

2021-09-10T19-47-09

Comment (Mothblocks) [reply code/__HELPERS/mobs.dm]: > I can't remember why it wasn't merged.

Because the last one didn't migrate savefiles--this one is only display names, and so requires no savefile changing.

I cleared it through oranges who was the one who closed the PR last time, and he said I could if I also showed the color next to it.

Comment (stylemistake) [reply code/__HELPERS/mobs.dm]: @Mothblocks well, how do I know which ones of them are supposed to be caucasian? It's all mixed up.

Comment (stylemistake) [reply code/__HELPERS/mobs.dm]: Also, this is olive: 
image

Comment (stylemistake) [reply code/__HELPERS/mobs.dm]: Anyway, not blocking the PR with this, but just saying that it would probably be better to keep the old names and sort them by increasing darkness, but grouped by ethnicity (which will avoid mixing up differences in hue).

Comment (Cyberboss) [reply code/__HELPERS/mobs.dm]: Forget one /s and everyone loses their minds

Comment (Mothblocks) [reply code/__HELPERS/mobs.dm]: image

I changed the names anyway to be more intuitive 😄

**Comment by oklews** [comment]
The lack of save button is REALLY offputting

**Comment by oklews** [comment]
the rest is great though!

**Comment by Guidesu** [comment]
Interface PreferencesMenu was not found
Can't even change my character

**Comment by Guidesu** [comment]
image

**Comment by Mothblocks** [comment]
I think that one's just from a cache fail, you might need to reset your BYOND cache.

**Review by Fikou — APPROVED** [review]
Code Looks Good

File: tgui/packages/tgui/interfaces/PreferencesMenu/KeybindingsPage.tsx
Comment (Watermelon914) [inline tgui/packages/tgui/interfaces/PreferencesMenu/KeybindingsPage.tsx:454]: you could probably shorten this code a lot by moving it to a functional component or using a for loop. A lot of it is the same.

**Comment by notfrying1pans** [comment]
image
I don't think broadcast login/logout should be under tooltips

**Comment by notfrying1pans** [comment]
image
It should say remove positive quirk here

File: code/modules/client/preferences/security_department.dm
Comment (AMonkeyThatCodes) [reply code/modules/client/preferences/security_department.dm:12]: pass in neither, just let it do it automatically.

File: tgui/packages/tgui/components/Dropdown.js
Comment (Mothblocks) [reply tgui/packages/tgui/components/Dropdown.js:9]: there's a few billion imports i don't use yeah, it's on my notion to remove them all

**Comment by TheSmallBlue** [comment]
Is it possible to make it so that rotating your character doesn't close the currently selected window? 

And this is kind of a stretch, but is it also possible to make it so that when you open a category it immediately scrolls down to your currently selected option? Like, if my character has a ponytail on, when i click the hairstyle button it immediately goes down to the ponytails.

**Comment by Mothblocks** [comment]
> Is it possible to make it so that rotating your character doesn't close the currently selected window?

Not a bad idea, though will be a bit tricky. I'll see what I can do.

**Comment by Mothblocks** [comment]
> And this is kind of a stretch, but is it also possible to make it so that when you open a category it immediately scrolls down to your currently selected option? Like, if my character has a ponytail on, when i click the hairstyle button it immediately goes down to the ponytails.

This is also doable, I'll see if I can get it in this version.

**Comment by Mothblocks** [comment]
@TheSmallBlue I think I'm going to push that off until this PR is merged--I don't want to hold up the whole codebase on anything other than bugs at this point.

**Comment by TheSmallBlue** [comment]
They were just suggestions, don't beat up yourself for not being able to do em. I don't mind.

File: code/modules/art/paintings.dm
Comment (Mothblocks) [reply code/modules/art/paintings.dm:339]: dont even need it to begin with since its always instantiated

**Review by AnturK — APPROVED** [review]
Looks good to me now. (Though we should probably think of some way to deal with all that preference specific initialize stuff later on).

File: code/modules/client/client_procs.dm
Comment (Watermelon914) [inline code/modules/client/client_procs.dm:1153]: change C to something more descriptive

File: code/modules/client/preferences/ghost.dm
Comment (Watermelon914) [inline code/modules/client/preferences/ghost.dm:172]: needs preferences here as second arg?

File: code/modules/client/client_procs.dm
Comment (Mothblocks) [reply code/modules/client/client_procs.dm:1153]: this is old, i dont even know why the hell this used client anyway considering its a verb on client

im changing it to src

File: code/modules/mob/living/carbon/human/species.dm
Comment (Watermelon914) [inline code/modules/mob/living/carbon/human/species.dm:251]: a lot of variables with 1 letter var names, would be nice if you could fix that

**Review by Watermelon914 — APPROVED** [review]

File: tgui/packages/tgui/styles/atomic/fit-text.scss
Comment (stylemistake) [inline tgui/packages/tgui/styles/atomic/fit-text.scss:8]: Indentation is 4, should be 2

File: tgui/packages/tgui/interfaces/PreferencesMenu/useRandomToggleState.ts
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/PreferencesMenu/useRandomToggleState.ts:6]: Is it not possible to just specify `useLocalState<boolean>` and get those types inferred?

Comment (Mothblocks) [reply tgui/packages/tgui/interfaces/PreferencesMenu/useRandomToggleState.ts:6]: It *can*, I'm just always a fan of explicitly contracting function arguments/return types.

Comment (stylemistake) [reply tgui/packages/tgui/interfaces/PreferencesMenu/useRandomToggleState.ts:6]: There is little point in doing that, as it would just be another point of breakage under refactoring.

**Comment by Yenwodyah** [comment]
Preferences menu randomly bluescreens and deletes all your settings now. Well done.

**Comment by Mothblocks** [comment]
It's not deleting anyone's settings, I've confirmed this a billion times. Be informed or get out of the way.

**Comment by Yenwodyah** [comment]
Stop fucking the game up plz

**Comment by LemonInTheDark** [comment]
No

**Comment by KittyNoodle** [comment]
ethereals for some reason get gendered pronouns, this seemed to line up with this merge

**Comment by Mothblocks** [comment]
Can you file a bug report, please? I wouldn't be surprised.

**Comment by Sparkezel** [comment]
image

File: code/modules/client/preferences.dm
Comment (Crossedfall) [inline code/modules/client/preferences.dm:1896]: This check wasn't ever re-added. #62571

**Pull request merged**
